Fix broken image region #5765

sstone-codaio opened the pull request:
Due to a refactoring change in #5721 we now return filename instead of filepath from editor.ImageUploadHandler, but in FilepathEditor.js the ajax callback is still expecting filepath and broke the image upload functionality.

Commit:
Fix broken image region #5765

Due to a refactoring change in #5721 we now return filename instead of filepath from editor.ImageUploadHandler, but in FilepathEditor.js the ajax callback is still expecting filepath and broke the image upload functionality.

seanlip commented:
LGTM, thanks! But: (a) did you test manually, and (b) are there other instances of filepath anywhere?

We should probably file an issue for adding e2e tests for this functionality.

Also passing review on to @nithusha21 @aks681 -- I'm currently trying to do a hotfix so am a bit distracted!

seanlip commented:
Also /cc @DubeySandeep

sstone-codaio commented:
Hi Sean
(a) yes I tested manually and 
(b) I doubly checked - no other instances of filepaths to my knowledge

nithusha21 commented:
Hi @sshou14 I tried by loading the test explorations (using the admin tab), but I still get the error message. Can you reproduce this? Or am I doing something wrong?

sstone-codaio commented:
Hi @nithusha21 what is the error you're getting, still page not found?

nithusha21 commented:
Yeah the error log is:

>   File "/Users/nithusha21/Documents/opensource/oppia/oppia_tools/google_appengine_1.9.67/google_appengine/lib/webapp2-2.5.2/webapp2.py", line 570, in dispatch
    return method(*args, **kwargs)
  File "/Users/nithusha21/Documents/opensource/oppia/oppia/core/domain/acl_decorators.py", line 41, in test_can_access
    return handler(self, *args, **kwargs)
  File "/Users/nithusha21/Documents/opensource/oppia/oppia/core/controllers/resources.py", line 79, in get
    raise self.PageNotFoundException
PageNotFoundException
WARNING  2018-10-19 06:16:23,555 base.py:492] Invalid URL requested: http://localhost:8181/assetsdevhandler/16/assets/image/code_sample_height_333_width_466.png

sstone-codaio commented:
Thank you. That's strange. Just to confirm I'm doing the same steps as you did, it was:
- Create an exploration.
- Add image region interaction.
- Upload an image.
- Click "Use this image",
- Check console.

Could you explain how to upload an image from the Admin tab? I can access the Admin dashboard but couldn't find an option to upload images.

nithusha21 commented:
I did the following:
- Login as admin (tick the checkbox when you login)
- Go to admin tab (from the dropdown when you hover over your profile image)
- reload exploration "three_balls" or "all_interactions"
- Go to the exploration (all_interactions is /explore/16)
- I see the error message.

I didn't try with a manual image upload. This was the first thing I tried, and that was giving the error message continuously.

seanlip commented:
Couple of random thoughts, off the top of my head:
- @nithusha21 does this happen if you check out the release-2.6.8 branch (say)?
- I remember #5501 but it looked like it was fixed. Not sure if related.

nithusha21 commented:
Tried on release-2.6.8 and it works fine!

DubeySandeep commented:
@sshou14 @nithusha21 For me it's working fine! (even uploading demo exploration)

nithusha21 commented:
@DubeySandeep on the PR branch or the release branch? The PR branch I am still running into issues. This might be just me though. Can anyone else repro this?

seanlip commented:
I can confirm that I see this too, on the PR branch. (Same error -- 404 with same file as @nithusha21.)

DubeySandeep commented:
@nithusha21, Sorry I missed the 4th step (Go to the exploration (all_interactions is /explore/16)), I was expecting error while reloading! 
I can also reproduce the issue! Thanks.

Commit:
Fix admin's exploration reload function

Due to refactoring we now use the same path for commiting image assets in dev and prod. But the exploration reload is still using the deprecated path for dev and therefore commiting images to the wrong path.

sstone-codaio commented:
Thanks for everyone's comment. All these are really helpful.

nithusha21 commented:
@sshou14 please note the backend test failure.

Commit:
Fix backend test failure

nithusha21 commented:
Hi @sshou14, the images are working fine now. However, this has broken audio for demo explorations (eg, try loading audio_test from the admin panel and go to /explore/17). I think it has something to do with the hardcoded "image" in the url, not sure.

nithusha21 commented:
Actually scratch that, the audio assets are also broken on develop. Are these related issues? If so I think they need to be fixed together. (Basically, the error messages are similar -- 404 on the asset URL)

sstone-codaio commented:
Thanks @nithusha21 you are right. The correct fix for the image assets would actually be adding image/ to their path. Right now we are using for audio paths like:
oppia/data/explorations/audio_test/assets/audio/hint-en-jca2lxjibh.mp3
but for image paths like:
oppia/data/explorations/solar_system/assets/Screen Shot 2015-02-18 at 3.08.44 AM.png

The hardcoded image was the wrong fix, instead we should make the second path:
oppia/data/explorations/solar_system/assets/image/Screen Shot 2015-02-18 at 3.08.44 AM.png

Commit:
Add image/ folder for demo assets to be consistent with audio assets

nithusha21 commented:
Hmmm, @sshou14 did you test audio manually? Audio assets still seem to be broken only on demo explorations though.

Could you manually test the following and make sure they work fine?
- Images load correctly when uploaded to an exploration
- Demo exploration images load correctly (try /explore/16 for example)
- Audio uploaded manually is playable (add audio using translation tab)
- Audio for demo explorations work correctly (try /explore/17 -- Audio test explorations, for example)

nithusha21 commented:
Hmmm, @sshou14 did you test audio manually? Audio assets still seem to be broken only on demo explorations though.

Could you manually test the following and make sure they work fine?
- Images load correctly when uploaded to an exploration
- Demo exploration images load correctly (try /explore/16 for example)
- Audio uploaded manually is playable (add audio using translation tab)
- Audio for demo explorations work correctly (try /explore/17 -- Audio test explorations, for example)

sstone-codaio commented:
Thanks @nithusha21  I tried the four combinations you mentioned and they all worked for me. Was the audio demo asset not working for you (/explore/17)?

seanlip reviewed (approved):
Changes LGTM code-wise. @nithusha21 could you please test and verify to make sure the bug you saw before is now fixed?

seanlip commented:
(Once that's done, let's merge this.)

nithusha21 commented:
Looks like I had to reload the exploration again from the admin page for it to work fine. I confirm that this fixes the asset related bug. Sorry for the confusion earlier! Thanks @sshou14 :)

sstone-codaio commented:
Thank you @nithusha21 !
